File: 020-valid-parentheses.py

class Solution:
    def isValid(self, s: str) -> bool:
        stack = list()
        left_paren = set("([{")
        right_paren = dict(zip(")]}","([{"))
        for ch in s:
            if ch in left_paren:
                stack.append(ch)
            elif ch in right_paren:
                if len(stack)==0 or stack[-1]!=right_paren[ch]:
                    return False
                stack.pop()
        return len(stack)==0

File: 020-valid-parentheses.cpp
#include <iostream>
#include <stack>
#include <string>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for(auto ch: s){
            if(ch=='['||ch=='{'){
                st.push(char(ch+2));
            }else if(ch=='('){
                st.push(')');
            }else if(st.empty()||st.top()!=ch){
                return false;
            }else{
                st.pop();
            }
        }
        return st.empty();
    }
};

int main(){
    Solution sol;
    string str;
    while(str!="q"){
        cin >> str;
        cout << sol.isValid(str) << "\n";
    }
    return 0;
}